Sources/GRPCCodeGen/Internal/Translator/IDLToStructuredSwiftTranslator.swift

```diff
@@ -112,87 +115,165 @@ extension IDLToStructuredSwiftTranslator {
+    let baseNames = service.methods.map { $0.name.base }
+    var seenBaseNames = Set<String>()
+
+    for baseName in baseNames {
+      if seenBaseNames.contains(baseName) {
+        throw CodeGenError(
+          code: .nonUniqueMethodName,
+          message: """
+            Methods of a service must have unique base names. \
+            \(baseName) is used as a base name for multiple methods of the \(service.name.base) service.
+            """
+        )
+      }
+      seenBaseNames.insert(baseName)
+    }
+
+    // Check that generated upper case names for methods are unique within a service, to ensure that
+    // the enums containing type aliases for each method of a service.
+    let upperCaseNames = service.methods.map { $0.name.generatedUpperCase }
+    var seenUpperCaseNames = Set<String>()
+
+    for upperCaseName in upperCaseNames {
+      if seenUpperCaseNames.contains(upperCaseName) {
+        throw CodeGenError(
+          code: .nonUniqueMethodName,
+          message: """
+            Methods of a service must have unique generated upper case names. \
+            \(upperCaseName) is used as a generated upper case name for multiple methods of the \(service.name.base) service.
+            """
+        )
+      }
+      seenUpperCaseNames.insert(upperCaseName)
+    }
+
+    // Check that generated lower case names for methods are unique within a service, to ensure that
+    // the function declarations and definitions from the same protocols and extensions have unique names.
+    let lowerCaseNames = service.methods.map { $0.name.generatedLowerCase }
+    var seenLowerCaseNames = Set<String>()
 
-    for methodName in methodNames {
-      if seenNames.contains(methodName) {
+    for lowerCaseName in lowerCaseNames {
+      if seenLowerCaseNames.contains(lowerCaseName) {
         throw CodeGenError(
           code: .nonUniqueMethodName,
           message: """
-            Methods of a service must have unique names. \
-            \(methodName) is used as a name for multiple methods of the \(service.name) service.
+            Methods of a service must have unique lower case names. \
+            \(lowerCaseName) is used as a signature name for multiple methods of the \(service.name.base) service.
             """
         )
       }
-      seenNames.insert(methodName)
+      seenLowerCaseNames.insert(lowerCaseName)
+    }
+  }
+
+  private func checkServiceDescriptorsAreUnique(
+    _ services: [CodeGenerationRequest.ServiceDescriptor]
+  ) throws {
+    let descriptors = services.map {
+      ($0.namespace.base.isEmpty ? "" : "\($0.namespace.base).") + $0.name.base
+    }
+    if let duplicate = descriptors.hasDuplicates() {
+      throw CodeGenError(
+        code: .nonUniqueServiceName,
+        message: """
+          Services must have unique descriptors. \
+          \(duplicate) is the descriptor of at least two different services.
+          """
+      )
     }
   }
```

Comment:
There's no need to create an intermediate array here, we can just do this directly which is a bit simpler:

```swift
var descriptors: Set<String> = []
for service in services {
  let name = service.namespace.base.isEmpty ? service.name.base : "\(service.namespace.base).\(service.name.base)"
  let (inserted, _) = descriptors.insert(name)
  if !inserted { 
    throw ...
  }
}
```